src/sc/sidechain.cpp

```diff
@@ -70,746 +66,87 @@ bool ScCoinsView::checkTxSemanticValidity(const CTransaction& tx, CValidationSta
-                            REJECT_INVALID, "sidechain-creation");
-                }
-            }
+            LogPrint("sc", "%s():%d - Invalid tx[%s] : sc creation withdrawalEpochLength %d is non-positive\n",
+                __func__, __LINE__, txHash.ToString(), sc.withdrawalEpochLength);
+            return state.DoS(100, error("%s: sc creation withdrawalEpochLength is not valid",
+                __func__), REJECT_INVALID, "sidechain-sc-creation-epoch-not-valid");
         }
-    }
-    return true;
-}
 
-bool ScCoinsView::IsTxApplicableToState(const CTransaction& tx)
-{
-    const uint256& txHash = tx.GetHash();
-
-    // check creation
-    BOOST_FOREACH(const auto& sc, tx.vsc_ccout)
-    {
-        const uint256& scId = sc.scId;
-        if (sidechainExists(scId))
+        if (!sc.CheckAmountRange(cumulatedAmount) )
         {
-            LogPrint("sc", "%s():%d - Invalid tx[%s] : scid[%s] already created\n",
-                __func__, __LINE__, txHash.ToString(), scId.ToString());
-            return false;
+            LogPrint("sc", "%s():%d - Invalid tx[%s] : sc creation amount is non-positive or larger than %s\n",
+                __func__, __LINE__, txHash.ToString(), FormatMoney(MAX_MONEY) );
+            return state.DoS(100, error("%s: sc creation amount is outside range",
+                __func__), REJECT_INVALID, "sidechain-sc-creation-amount-outside-range");
         }
-        LogPrint("sc", "%s():%d - OK: tx[%s] is creating scId[%s]\n",
-            __func__, __LINE__, txHash.ToString(), scId.ToString());
     }
 
-    // check fw tx
-    BOOST_FOREACH(const auto& ft, tx.vft_ccout)
+    for (const auto& ft : tx.GetVftCcOut())
     {
-        const uint256& scId = ft.scId;
-        if (!sidechainExists(scId))
+        if (!ft.CheckAmountRange(cumulatedAmount) )
         {
-            // return error unless we are creating this sc in the current tx
-            if (!hasScCreationOutput(tx, scId) )
-            {
-                LogPrint("sc", "%s():%d - tx[%s] tries to send funds to scId[%s] not yet created\n",
-                    __func__, __LINE__, txHash.ToString(), scId.ToString() );
-                return false;
-            }
+            LogPrint("sc", "%s():%d - Invalid tx[%s] : sc creation amount is non-positive or larger than %s\n",
+                __func__, __LINE__, txHash.ToString(), FormatMoney(MAX_MONEY) );
+            return state.DoS(100, error("%s: sc creation amount is outside range",
+                __func__), REJECT_INVALID, "sidechain-sc-creation-amount-outside-range");
         }
-        LogPrint("sc", "%s():%d - OK: tx[%s] is sending [%s] to scId[%s]\n",
-            __func__, __LINE__, txHash.ToString(), FormatMoney(ft.nValue), scId.ToString());
     }
+
     return true;
 }
 
-bool ScCoinsView::hasScCreationOutput(const CTransaction& tx, const uint256& scId)
+bool Sidechain::anyForwardTransaction(const CTransaction& tx, const uint256& scId)
```

Comment:
Do we need this method? I can't see the usage of it.
Also the name is not clear to me. What about the name `hasForwardTransferOutput`?